app/views/unattended/provisioning_templates/PXELinux/preseed_default_pxelinux.erb

```diff
@@ -31,8 +31,8 @@ test_on:
   else
     options << 'console-setup/ask_detect=false console-setup/layout=USA console-setup/variant=USA keyboard-configuration/layoutcode=us localechooser/translation/warn-light=true localechooser/translation/warn-severe=true'
   end
-  if @host.provision_interface.vlanid.present?
-    options << "netcfg/use_vlan=true netcfg/vlan_id=#{@host.provision_interface.vlanid}"
+  if @host.provision_interface.virtual? && @host.provision_interface.tag.present?
+    options << "netcfg/use_vlan=true netcfg/vlan_id=#{@host.provision_interface.tag}"
```

Comment:
Hey, thanks for the patch.

Please use `vlanid`, the method will use `tag` if defined, subnet vlanid otherwise.

```suggestion
    options << "netcfg/use_vlan=true netcfg/vlan_id=#{@host.provision_interface.vlanid}"
```

Reply:
I don't know, if that's a good idea. I just want to use the vlan tag if its a virtual interface and the tag has been set. If that's the case the vlanid function will always return the tag, anyway, so why going through the function? Because beyond returning the tag the function does only things I specifically do not want here.

Reply:
I don't understand, the method returns empty string if no tag is set on NIC or Subnet(s). And

```
"".present:
=> false
```

Reply:
> I don't understand, the method returns empty string if no tag is set on NIC or Subnet(s). And
> 
> ```
> "".present?
> => false
> ```

I don't understand, what you don't understand. If there is no tag (aka false), then don't configure vlan tagging.

Reply:
Let me try it this way: users can define VLAN tags not only on NIC, but also on Subnet. This patch ignores that feature and renders it unused (granted it has been broken in templates for some time now).

Reply:
> Let me try it this way: users can define VLAN tags not only on NIC, but also on Subnet. This patch ignores that feature and renders it unused (granted it has been broken in templates for some time now).

Yes. But only because a subnet has a vlan tag assigned (because it runs tagged in your network), doesn't mean that your host interface must also be tagged.

If I connect my vlan untagged to the host, I must not configure the vlan tagging on the provision interface.

Only if I connect the vlan tagged to the host, I must configure vlan tagged on the provisioning interface.

Both is possible at the same time on different hosts and thus must be distinguished.

Reply:
In the end, it all nails down to the question, when exactly the "virtual" flag is set on an interface. I don't find it in the code where it is set exactly. To get this properly working, the condition must check if this interface has been configured to be tagged or not (regardless whether the associated subnet is tagged in the network and has a tag configured). So far, I have been looking at the data I can access in the template and guessed what is set when.

Reply:
So you have one subnet but it is available both via tagged and untagged frames? How this is implemented? I am just curious.

Well, I was thinking to create two subnets, one tagged, other not tagged. But I realized foreman does not allow creation of two same subnets (with or without tag). So what you say makes sense, it looks like VLAN tag on subnet is not very useful. It looks more informative.

In network configuration scripts we have the following:

```erb
<%- if @interface.virtual? && (!@subnet.nil? && (@subnet.has_vlanid? || @interface.vlanid.present?)) -%>
<%=   "VLAN=yes" %>
<%-   if @attached_to_bond -%>
<%-     if ( @host.operatingsystem.family == 'Redhat' && @host.operatingsystem.major.to_i <= 7 ) -%>
<%=       "NM_CONTROLLED=no" %>
<%-     end -%>
<%-   end -%>
<%-   if @interface.identifier.match(/^vlan\d+/) -%>
<%=     "VLAN_NAME_TYPE=VLAN_PLUS_VID_NO_PAD" %>
<%=     "PHYSDEV=#{@interface.attached_to}" %>
<%-     if @bonding_interfaces.include?(@interface.attached_to) -%>
<%=       "TYPE=bonding" %>
<%-     end -%>
<%-   end -%>
```

So it drops VLAN=yes if interface is virtual and subnet has a tag or interface has a tag. The `has_vlanid?` is probably not needed as it is covered by the `vlanid` method.

@ares do you remember how this was designed? What was the purpose of VLAN tag on a subnet?

I think this is the opportunity to standardize this across templates and perhaps drop some info into documentation about this. If we choose to change this how you propose, let's do it everywhere.

Reply:
> So you have one subnet but it is available both via tagged and untagged frames? How this is implemented? I am just curious.

That is standard switch configuration. You configure the switch port to the host in "access mode" and assign it to a VLAN. For the host it's a standard untagged network connection and it's completely oblivious of the VLAN it belongs to.

Isn't that even the standard way for VMs? The network switch connects all necessary VLANs tagged to the hypervisor and then you connect one VLAN to one VM network port which is also usually handled untagged.

And if you want, you can also connect additional VLANs as tagged on the same interface, e.g. have VLAN 10 untagged and VLAN 20 tagged. One advantage of the untagged vlan is that you can easily change it on the switch while for tagged vlans you'll have to modified the VLAN tag information on the server and the switch accordingly.

> So what you say makes sense, it looks like VLAN tag on subnet is not very useful. It looks more informative.

At the moment, it is. It's very useful for me to know what VLAN ID is associated with a specific subnet. It could be more useful, if that information was used to prefill the tag field for a interface.

> So it drops VLAN=yes if interface is virtual and subnet has a tag or interface has a tag. The `has_vlanid?` is probably not needed as it is covered by the `vlanid` method.

The problem with this is, that a bond interface is also marked as "virtual". That I know from reverse engineering and that's what I have asked when exactly an interface is considered "virtual" and where the code is which sets this "virtual" flag on an interface. I know of bond interfaces and normal interface where I select the virtual checkbox in the gui.

We have quite a few hardware server with redundant, LACP bonded ethernet interfaces which usually have one VLAN untagged and sometimes additional VLANs tagged. And this kind of scenario is exactly causing my problems like with the script you have posted. The interface is virtual because it's a bond, my subnet has a vlan tag set but for the actual host interface it's connected untagged.

Of course, the script would only set up tagging if the interface identifier is /vlan\d+/ So it would work correctly if my interface was bond0 and is untagged. Of course, vlan alias names are also quite common and what we usually use. So with that script, I would have a problem if I wanted to use bond0:10 to get a tagged connection to VLAN 10 on bond0.

Either way, technically, for me a subnet can have a vlan tag associated but for the host interface it depends on the specific host connection. That's why it's a host setting for me.

Ideally for me, would be if the GUI would allow you to choose the subnet as it is and then have a simple checkbox for each interface allow you to enable/disable tagging on that specific interface. By default, this would enable tagging with the tag defined in the subnet. To make it perfect, an override option for the tag would be useful incase you are in a network where you do some evil vlan tag mappings, thus changing tags for the same subnet depending on where you are.

But well, that's for the future. At the moment there is no simply switch to turn on/off tagging thus it's necessary to find a condition which covers this decision. As far as I can tell, the interface must be marked "virtual" but with a  bond interface being virtual, too, there are some exceptions. I can specifically exclude bond interfaces from the condition but I'd rather do it right. And reverse engineering only gets me so far...